src/controller/AudioController.vala

```diff
@@ -0,0 +1,99 @@
+/*
+ * brickman -- Brick Manager for LEGO MINDSTORMS EV3/ev3dev
+ *
+ * Copyright 2014-2015 David Lechner <david@lechnology.com>
+ *
+ * This program is free software; you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation; either version 2 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston,
+ * MA 02110-1301, USA.
+ */
+
+/* BatteryController.vala - Controller for monitoring battery */
+
+using Ev3devKit.Devices;
+using Ev3devKit.Ui;
+using Alsa;
+
+namespace BrickManager {
+    public class AudioController : Object, IBrickManagerModule {
+        private const int VOLUME_STEP = 10;
+
+        Mixer mixer;
+        MixerElementSelectorWindow mixer_select_window;
+        MixerElementVolumeWindow volume_window;
+
+        public string display_name { get { return "Sound"; } }
+
+        protected void initialize_mixer() {
+            mixer = null;
+            Mixer.open(out mixer);
+            mixer.attach();
+            mixer.register();
+            mixer.load();
+        }
+
+        void create_main_window () {
+            mixer_select_window = new MixerElementSelectorWindow ();
+
+            mixer_select_window.mixer_elem_selected.connect ((selected_element) => {
+                if(volume_window == null)
+                    create_volume_window();
+
+                volume_window.current_element = selected_element;
+                volume_window.show_element_details = true;
+                volume_window.show();
+            });
+        }
+
+        void create_volume_window() {
+            volume_window = new MixerElementVolumeWindow();
+
+            // Wire up handlers for volume window signals
+            volume_window.volume_up.connect(() =>
+                volume_window.current_element.volume += VOLUME_STEP);
```

Comment:
A reference cycle leads to a memory leak. https://wiki.gnome.org/Projects/Vala/ReferenceHandling

So what happens here is that in the generated C code, it takes a reference to `volume_window` in the lambda expression and creates a closure structure. That closure is assigned to `volume_window` when the signal is connected, so now `volume_window` has a reference to itself and will never be freed.

So, you have to break this reference cycle. You could do this by disconnecting the signal, for example when the window is closed.

You can also do it by using a weak reference to `volume_window`. This way, when the C closure is created, it does not take a reference to `volume_window`. Since the signal will be disconnected when `volume_window` is destroyed, you don't have to worry about the signal being called after `volume_window` is freed.